src/core/utils/components/VirtualScroll.tsx

```diff
@@ -100,145 +131,232 @@ export type VirtualScrollProps = {
+    // because before that calculations are not right
+    const [isMounted, setIsMounted] = React.useState(false);
 
-    resizeRef(scrollableParent);
-  }, [resizeRef]);
+    const onResize = React.useCallback(({ height }) => {
+      if (height > 0) {
+        setIsMounted(true);
+      }
+      setScrollContainerHeight(height);
+    }, []);
+    const [resizeRef, resizeObserver] = useResizeObserver(onResize);
+    const Component = as;
 
-  const visibleChildren = React.useMemo(() => {
-    const arr = [];
-    const endIndex = Math.min(
-      itemsLength,
-      startNode + visibleNodeCount + bufferSize * 2,
-    );
-    for (let i = startNode; i < endIndex; i++) {
-      arr.push(itemRenderer(i));
-    }
-    return arr;
-  }, [itemsLength, itemRenderer, bufferSize, startNode, visibleNodeCount]);
-
-  // Get child height when children available
-  React.useLayoutEffect(() => {
-    if (!parentRef.current || !visibleChildren.length) {
-      return;
-    }
-
-    const firstChild = parentRef.current.children.item(0) as HTMLElement;
-    childHeight.current = Number(getElementHeight(firstChild).toFixed(2));
-  }, [visibleChildren.length]);
-
-  const updateVirtualScroll = React.useCallback(() => {
-    const scrollableContainer =
+    // Find scrollable parent
+    // Needed only on init
+    React.useLayoutEffect(() => {
+      const scrollableParent = getScrollableParent(
+        parentRef.current,
+        parentRef.current?.ownerDocument,
+      );
+      scrollContainer.current = scrollableParent;
+
+      resizeRef(scrollableParent);
+    }, [resizeRef]);
+
+    // Stop watching resize, when virtual scroll is unmounted
+    React.useLayoutEffect(() => {
+      return () => resizeObserver?.disconnect();
+    }, [resizeObserver]);
+
+    const getScrollableContainer = () =>
       scrollContainer.current ??
       (parentRef.current?.ownerDocument.scrollingElement as HTMLElement);
-    if (!scrollableContainer) {
-      return;
-    }
-    const start = getNumberOfNodesInHeight(
-      childHeight.current,
-      scrollableContainer.scrollTop,
-    );
-    const startIndex = Math.max(0, start - bufferSize);
-    setStartNode(startIndex);
-    setVisibleNodeCount(
-      getVisibleNodeCount(
-        childHeight.current,
+
+    const visibleChildren = React.useMemo(() => {
+      const arr = [];
+      const endIndex = Math.min(
+        itemsLength,
+        startNode + visibleNodeCount + bufferSize * 2,
+      );
+      for (let i = startNode; i < endIndex; i++) {
+        arr.push(itemRenderer(i));
+      }
+      return arr;
+    }, [itemsLength, itemRenderer, bufferSize, startNode, visibleNodeCount]);
+
+    // Get child height when children available
+    React.useLayoutEffect(() => {
+      if (!parentRef.current || !visibleChildren.length) {
+        return;
+      }
+
+      const firstChild = parentRef.current.children.item(0) as HTMLElement;
+      const child = parentRef.current.children.item(1) as HTMLElement;
```

Comment:
nit: maybe rename this to `secondChild`. I got confused with all the different children in here

Reply:
renamed